### ISSUE TITLE
Serializing monads

### ISSUE THREAD

**Issue by MarcNq** [true-myth/true-myth#48]
Hi guys,

I have an unorthodox but really easy request. I hope you will find it reasonable enough.

It would be great if you could **provide constants to access the data internals** of Maybe and Result types (`.value` and `.error` properties).

Alternatively, you could **add to the specs those data internals**, so that if they change in the future, you consider it a breaking change and a SEMVER major release.

The reason is we need to deserialize monads and to do so we need to be coupled to these internals.

If you want more detail:

## Context

We have a client with Redux that stores the app state in local storage, thus serializing all state. We want to store monads in the state. We know it is a bad practice to store classes in Redux, but Maybes and Results are data types that hold state, and they make much sense to be stored.

The Maybe and Result types serialize correctly to the following examples:

```javascript
{variant: 'Just', value: 4}
{variant: 'Nothing'}
{variant: 'Ok', value: 4}
{variant: 'Err', error: 'hi!'}
```

We make our reducers detect the de-serialization action, and re-create the monads from the objects using `Maybe.just(...)`, `Maybe.nothing()`, and so on.

## The Problem

In order to de-serialize the monads, **we need to be coupled to the object's internals**: `.variant`, `.value`, and `.error`.

The Variant property is well specified in your docs, and in fact we are using the lib's constants `Variant.Just`, `Variant.Nothing`, etc. But the other internals are not specified (because they are internals), and as such are subject to possible change without notice.

If this change causes a major release, it allows this coupling because it is then our responsibility to keep our code consistent with the update.
If these internals are exposed through a constant, then you can change them without provoking a breaking change.

## Why not a deserialize method

Deserializing involves type refinement from input source, it is a whole other field in itself, and it can devolve into libraries as complex as [io-ts](https://github.com/gcanti/io-ts), so in my humble opinion, it doesn't make sense as a **true-myth** responsibility, but belonging to userland.

Thank you for your attention!

**Comment by chriskrycho** [comment]
Thanks for the great writeup! This is totally a use case we should support.

The obvious alternative that occurs to me is, instead of making those invariants hold (in case we find a better/faster/lighter-weight way to do this in the future) is for us to supply a `toJSON()` method so that `JSON.stringify()` does the right thing, supplying sufficient information for serialization and deserialization.

```ts
class Just<T> {
  toJSON() {
    const { variant, value } = this;
    return { variant, value };
  }
}

class Nothing<T> {
  toJSON() {
    return { variant: this.variant };
  }
}
```

(And the same for `Result`.)

I *think* that would solve this in a way that would give nice public API that Just Works™ for the normal serialization/deserialization approach, while leaving private internals private. Thoughts?

**Comment by MarcNq** [comment]
Thank *you* for the swift response!

That would keep it decoupled and consistent indeed.

**Comment by chriskrycho** [comment]
If you want to open a PR with it, we’ll merge it; otherwise, I’ll try to get to it sometime late this week! (It’s super small; I’m just particularly busy at the moment!)

**Comment by MarcNq** [comment]
Please let me open a PR for it. I may need some help with the docs, but since this is due to our needs, I think it would be ungrateful from us otherwise.

**Comment by chriskrycho** [comment]
Very good—and we're happy to help provide guidance and iteration on the docs! And no worries regardless; I'm actually surprised this is the first time this has come up.

**Comment by MarcNq** [comment]
Upon closer inspection, I think we should also implement [`valueOf`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/valueOf) so that the object is consistently coerced to primitive types.

I would make `toJSON` return the result of `valueOf(this)`.
Does that make sense to you?

Credit for the idea goes to this [tweet by Joelnet](https://twitter.com/joelnet/status/1158851253545271296?s=19).

**Comment by CrossEye** [comment]
Implementing `valueOf` suggests that there is an inherent order to a type, that there is a primitive value specifically associated with each member.  Is that what's really wanted here?  `toString` and `toJSON` make perfect sense to me, but this is less clear.

That tweet was in reference to a custom numeric type, which has an obvious primitive representation.

**Comment by MarcNq** [comment]
Well, _there is_ a primitive value specifically associated with each monad member, because `valueOf` already exists for any JS object. Right now, though, JS will be using the default implementation, which may leak implementation details and be inconsistent with `toJSON`.

By implementing `valueOf`, we can control this behavior.

**Comment by chriskrycho** [comment]
Quoting my comments in #49:

> I decided against using `valueOf` *at all* because I think it's important to let the owners of the types passed in be in control of any of that kind of serialization representation. `JSON.stringify` is smart and does the right thing by recursively using `toJSON` (and `toValue`) on types as supplied and/or appropriate. What's more, while at the end of a `JSON.stringify(someMaybe)` everything will end up being a string, that's intentionally not true of `someMaybe.toJSON()` invocation, and in fact users might have other formats besides JSON strings they'd like to target (protobuf, etc.). By actively making a decision, we'd *prevent* users from making their own decisions on serialization of their data, which is just a no-go.

**Comment by MarcNq** [comment]
Thank you for your review! I just came back from holidays, I'll update the PR today if possible.

**Comment by MarcNq** [comment]
Hi again! I agree with all the changes you requested, and I have applied them (not only the inline suggestions). Check them when possible.
Thanks for your patience!

**Comment by chriskrycho** [comment]
Released in 3.1.0! Thanks again!
